library/psa_crypto.c

```diff
@@ -4411,40 +4414,49 @@ static psa_status_t psa_key_derivation_hkdf_read( psa_hkdf_key_derivation_t *hkd
         if( hkdf->block_number == 0xff )
             return( PSA_ERROR_BAD_STATE );
 
+
+        if( PSA_ALG_IS_HKDF_EXTRACT( kdf_alg ) && hkdf->block_number == 0 )
```

Comment:
Ok, I think this is correct, and the rest of the changes in this function as well, but I find this function pretty hard to reason about it is current state, because it's just doing very different things depending on whether the alg is HKDF-Extract on one hand, or HKDF / HKDF-Expand on the other hand. So different in fact, that I think it would be better to just have two separate functions here, and dispatch to the right one in `psa_key_derivation_output_bytes()`. I think the function for HKDF-Extract would be very simple as it can copy directly from `prk` and doesn't need to care about the case the `ouput_length` is more than one block.

It's all too easy when adding new features to existing code to slowly increase the complexity and decrease the readability, so let's pay extra attention to that!

Reply:
I agree that it became more complex, but I think this is the easiest way to handle reading key in parts.
To be honest I don't understand why we allow reading key in parts, but it works like this for other algorithms so it should also work like this for reading `prk` (`HKDF-Extract`).
I can add function `psa_key_derivation_hkdf_extract_read()`, but the loop that allows reading parts of key will be duplicated.

Reply:
Thanks for trying it out! Agreed, using the same field, in this case `hkdf->offset_in_block`, for two different things doesn't sound great either. I'll think about it.

> To be honest I don't understand why we allow reading key in parts, but it works like this for other algorithms so it should also work like this for reading prk (HKDF-Extract).

Yes, for other KDFs it is useful to be able to read the output in parts, for example TLS 1.2 uses its PRF by drawing n bits for the encryption key, n bits for the decryption key, etc. For HKDF-Extract I don't think it ever makes sense to read the output in parts, but it should probably work like other algorithms for the sake of uniformity.

Reply:
Ok, I tried writing an alternative proposal for how to handle this, you can find it here: the [branch](https://github.com/mpg/mbedtls/tree/5834-prop), the [commit](https://github.com/mpg/mbedtls/commit/3676419d5813d417ad216a385bdb1c773a0436cc). Compared to your version, it does more in `hkdf_input()` and less in `hkdf_read()`. IMO that reflects the difference between Extract and the other two: with Extract the output is pre-computed when reading the key, while with HDKF(-Expand) it's computed on-demand when reading it.

It's of course debatable, and partially a matter of taste, which version is better. But I think I can identify two specific points that bugged me when first reading the version in the present PR:

1. The `memcpy()` from `prk` to `output_block` is visually inside a while loop, but only happens once. IMO it should ideally not look like it's in a loop, or at least there should be a comment explaining the flow (the loop will run twice: the first time generating no output, then doing the memcpy, then it will generate output then stop because we know we're not being asked for more than one block of output).
2. The limit we check is always 0xff even for Extract. This is not just readability: if we have a defensive check it should work for all cases, but I think for me it impacted readability as well, making it harder to understand how many times the loop will run.

I'm not necessarily pushing for you to adopt my proposal, but whatever solution you come up with, I'd like it to address the above points.

Reply:
(Posted the comment above just before I saw you pushed new commits, will have a look at them now.)

Reply:
Reviewed the proposal and it looks good to me. I like the idea to copy `prk` to `output` immediately after it was generated. This simplifies `hkdf_read()`. I will adapt code.
Thanks!

Reply:
Cool, I'm happy we reached a quick agreement on that one!